pkgs/pkgs-lib/formats.nix

```diff
@@ -60,14 +65,20 @@ rec {
 
   };
 
-  yaml = {}: {
+  yaml = {
+    comment ? "Generated with Nix",
+    ...
+  }: {
 
-    generate = name: value: pkgs.callPackage ({ runCommand, remarshal }: runCommand name {
+    generate = let
+      addComment = lib.optionalString (comment != null) "echo ${lib.escapeShellArg (addCommentPrefix "#" comment)} > $out";
```

Comment:
Instead of doing this at evaluation time, I think it would be better to push it to build time inside the derivation, this way we don't increase evaluation time. In `lib` code this wouldn't be possible because you can't use `pkgs`, but here we can!

To do this, just pass `inherit comment` to the derivation arguments (might need some `null` handling though), set `passAsFile = [ "comment" ]`, then use `${pkgs.moreutils}/bin/ts '# ' "$commentFile" > $out` to do the prepending of `# `.

And side benefit: The comment can refer to derivations without causing IFD.

Same for the other formats if possible. In fact it would be ideal if all computation happens at build time, so that these generators are mostly free to use regarding evaluation time.

Reply:
I'll look into it!